source/extensions/filters/http/common/compressor/compressor.cc

```diff
@@ -0,0 +1,354 @@
+                          "application/json", "image/svg+xml", "text/xml",
+                          "application/xhtml+xml"});
+}
+
+} // namespace
+
+CompressorFilterConfig::CompressorFilterConfig(
+    const envoy::config::filter::http::compressor::v2::Compressor& compressor,
+    const std::string& stats_prefix, Stats::Scope& scope, Runtime::Loader& runtime,
+    const std::string& content_encoding)
+    : compressor_registry_(compressorRegistry()),
+      content_length_(contentLengthUint(compressor.content_length().value())),
+      content_type_values_(contentTypeSet(compressor.content_type())),
+      disable_on_etag_header_(compressor.disable_on_etag_header()),
+      remove_accept_encoding_header_(compressor.remove_accept_encoding_header()),
+      stats_(generateStats(stats_prefix, scope)), runtime_(runtime),
+      content_encoding_(content_encoding) {
+  Thread::LockGuard lock(compressor_registry_.mutex_);
+  compressor_registry_.compressors_.insert({this, compressor_registry_.registration_count_});
+  compressor_registry_.registration_count_++;
+}
+
+CompressorFilterConfig::~CompressorFilterConfig() {
+  Thread::LockGuard lock(compressor_registry_.mutex_);
+  ASSERT(compressor_registry_.compressors_.count(this) > 0);
+  compressor_registry_.compressors_.erase(this);
+}
+
+StringUtil::CaseUnorderedSet
+CompressorFilterConfig::contentTypeSet(const Protobuf::RepeatedPtrField<std::string>& types) {
+  const auto& default_content_encodings = defaultContentEncoding();
+  return types.empty() ? StringUtil::CaseUnorderedSet(default_content_encodings.begin(),
+                                                      default_content_encodings.end())
+                       : StringUtil::CaseUnorderedSet(types.cbegin(), types.cend());
+}
+
+uint32_t CompressorFilterConfig::contentLengthUint(Protobuf::uint32 length) {
+  return length > 0 ? length : DefaultMinimumContentLength;
+}
+
+CompressorFilterConfig::CompressorRegistry& CompressorFilterConfig::compressorRegistry() {
+  MUTABLE_CONSTRUCT_ON_FIRST_USE(CompressorRegistry);
+}
+
+const std::map<std::string, uint32_t> CompressorFilterConfig::registeredCompressors() const {
+  std::map<std::string, uint32_t> encodings;
+  std::vector<std::pair<std::string, uint32_t>> temp;
+
+  {
+    // Make a temp copy to release the lock earlier.
+    Thread::LockGuard lock(compressor_registry_.mutex_);
+    for (const auto& item : compressor_registry_.compressors_) {
+      temp.push_back({item.first->contentEncoding(), item.second});
+    }
+  }
+
+  // There could be many compressors registered for the same content encoding, e.g. consider a case
+  // when there are two gzip filters using different compression levels for different content sizes.
+  // In such case we ignore duplicates (or different filters for the same encoding) registered last.
+  for (const auto& item : temp) {
+    if (encodings.count(item.first)) {
```

Comment:
do the map lookup once rather than twice. ie capture the iterator in a temp with a find(), then compare it to .end() to replace count() as an if-condition. Then dereference iter->second rather than calling at().